### ISSUE TITLE
OpenSSLCertificateOptions doesn’t disallow SSLv2

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6337]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)| @hynek reported|
|-|-|
|Trac ID|trac#6337|
|Type|defect|
|Created|2013-02-28 13:29:15Z|
|Branch|https://github.com/twisted/twisted/tree/disable-sslv2-6337|

Our default SSL method in `DefaultOpenSSLContextFactory` and thus in string endpoints is `SSLv23` for maximum compatibility, but we disable `SSLv2` because it’s insecure.

However, that doesn’t happen in `OpenSSLCertificateOptions`: http://twistedmatrix.com/trac/browser/trunk/twisted/internet/_sslverify.py#L743

Probably because it is newer and uses `TLSv1` as default?

I suggest to set `OP_NO_SSLv2` here too by default. If that’s not possible, at least make it possible to inject it from the outside by a new option. Otherwise we can’t switch string endpoints to `OpenSSLCertificateOptions` in [#6286](https://github.com/twisted/twisted/issues/6286).

Attachments:

* [disable-sslv2-6337.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ba/6baf602fca3995ed3bb62c44dec6adb301140add/829638f12b1004b8b1775d0d025c2a7a2d6c9460.patch) (3877 bytes) - added by hynek on 2013-03-02 16:48:21Z - 
* [disable-sslv2-6337-v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ba/6baf602fca3995ed3bb62c44dec6adb301140add/d6be1323ac3fc73c950f25fded71aa3ce66181da.patch) (4387 bytes) - added by hynek on 2013-03-02 21:10:32Z - Better topfile.

<details><summary>Searchable metadata</summary>

```
trac-id__6337 6337
type__defect defect
reporter__hynek hynek
priority__normal normal
milestone__ 
branch__branches_disable_sslv2_6337 branches/disable-sslv2-6337
branch_author__hynek hynek
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1362058155000000 1362058155000000
changetime__1362605849000000 1362605849000000
version__None None
owner__tom_prince tom.prince
cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

It should be set by default if someone chooses SSLv23. For other methods it's either irrelevant, or if someone explicitly asks for the SSLv2 method they should be able to get SSLv2. No reason to have external option, SSLv2 is ancient.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

You may be interested to know that most distributors of OpenSSL (Linux vendors, largely) have taken this decision out of our hands already.  At least a year ago they began building OpenSSL without SSLv2 support.  On such builds, SSLv23 does not enable SSLv2, only SSLv3 and TLSv1.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

I believe that the right thing to do is to _always_ use SSLv23 because that means "more than one specific method" and then use the OP_NO_WHATEVER flags to turn off older protocol versions.  As newer TLS versions become available, setting a specific method (as `OpenSSLCertificateOptions` currently does) might, depending on the whims of the OpenSSL maintainers, mean that we don't support newer, _more_ secure TLS versions, even if the library could do that for us.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_5"></a>|@hynek commented|
|-|-|

IOW, we should make `OpenSSLCertificateOptions`’s default identical to `DefaultOpenSSLContextFactory` and use SSLv23 with disabled SSLv2 as default?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Replying to [hynek](#note_5):
> IOW, we should make `OpenSSLCertificateOptions`’s default identical to `DefaultOpenSSLContextFactory` and use SSLv23 with disabled SSLv2 as default?

No; we already require TLS 1.0, so apparently disabling SSLv3 by default is fine too.  [TLS 1.0 was defined in 1999](https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_1.0) so most applications are caught up by now.

I think it would be better to give users a way to express a 'minimum version', and then set all apropriate OP_NO_WHATEVER options, rather than let them set whatever method and then try to bulletproof it for them.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_7"></a>|@hynek commented|
|-|-|

I see. 

You are basically saying, we should abstract the method selection away from OpenSSL’s crazy modus operandi as per the discussion on IRC yesterday, right?

That sounds like an extra ticket to me… Would you be cool with fixing this by behaving like `DefaultOpenSSLContextFactory` for now because I *need* a way to specify SSLv23 w/o SSLv2 for [#6286](https://github.com/twisted/twisted/issues/6286) if we don’t want to make TLSv1 default for str ports?

As for the extra SSL/TLS work I wouldn’t mind to help out after my chains are finally merged. :)  I’d be happy to discuss the general future of SSL in Twisted at PyCon too.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Replying to [hynek](#note_7):
> I see. 
> 
> You are basically saying, we should abstract the method selection away from OpenSSL’s crazy modus operandi as per the discussion on IRC yesterday, right?

Right.  This would also make it easier to implement with other libraries.

> That sounds like an extra ticket to me… Would you be cool with fixing this by behaving like `DefaultOpenSSLContextFactory` for now because I *need* a way to specify SSLv23 w/o SSLv2 for [#6286](https://github.com/twisted/twisted/issues/6286) if we don’t want to make TLSv1 default for str ports?

Aah, I see.  You want to leave it that way so it will be a compatible change, and not change the default version for strports.

I would rather do the minimum-version thing in this ticket.  It's really not much work; since you realistically only have 3 options for minimum version; SSLv2 (which hopefully is not even available in your build of OpenSSL; at any rate it shouldn't be), SSLv3, and TLSv1.  Which means you can either set OP_NO_SSLv2 or OP_NO_SSLv3... and that's it.  One extra conditional does not seem like an onerous additional amount of code :).

> As for the extra SSL/TLS work I wouldn’t mind to help out after my chains are finally merged. :)  I’d be happy to discuss the general future of SSL in Twisted at PyCon too.

I would _love_ to do that.  I'll look forward to seeing you there.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_9"></a>|@hynek set owner to @hynek|
|-|-|

Replying to [glyph](#note_8):

> > That sounds like an extra ticket to me… Would you be cool with fixing this by behaving like `DefaultOpenSSLContextFactory` for now because I *need* a way to specify SSLv23 w/o SSLv2 for [[#6286](https://github.com/twisted/twisted/issues/6286)](https://github.com/twisted/twisted/issues/6286) if we don’t want to make TLSv1 default for str ports?
> Aah, I see.  You want to leave it that way so it will be a compatible change, and not change the default version for strports.

Exactly. That’s currently blocking #6286 from merging.

> I would rather do the minimum-version thing in this ticket.  It's really not much work; since you realistically only have 3 options for minimum version; SSLv2 (which hopefully is not even available in your build of OpenSSL; at any rate it shouldn't be), SSLv3, and TLSv1.  Which means you can either set OP_NO_SSLv2 or OP_NO_SSLv3... and that's it.  One extra conditional does not seem like an onerous additional amount of code :).

No it isn’t. I just tend to shy away from adding options. :) Will do.
 
> > As for the extra SSL/TLS work I wouldn’t mind to help out after my chains are finally merged. :)  I’d be happy to discuss the general future of SSL in Twisted at PyCon too.
> I would _love_ to do that.  I'll look forward to seeing you there.

Awesome!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

I think most of what's written in the comments above is good and interesting.  However, if you just want to resolve this ticket, then it's an absolutely trivialy change:

```
    + ctx.set_options(SSL.OP_NO_SSLv2)
```

(plus unit tests, obviously).  Since this ticket is blocking other more interesting work, I think it's worth considering doing this the simply way and deferring the rest of these improvements to a later ticket.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_10):
> I think most of what's written in the comments above is good and interesting.  However, if you just want to resolve this ticket, then it's an absolutely trivialy change:
> 
> ```
>     + ctx.set_options(SSL.OP_NO_SSLv2)
> ```
> 
> (plus unit tests, obviously).  Since this ticket is blocking other more interesting work, I think it's worth considering doing this the simply way and deferring the rest of these improvements to a later ticket.

I wish I could agree, I like trivial changes. But if upgrading the protocol version for users of SSL endpoints to TLSv1 is an incompatible change, then wouldn't upgrading the protocol version for `OpenSSLCertificateOptions` users who explicitly specified `SSLv23_METHOD` to SSL version 3 _also_ be an incompatible change?  (Perhaps a worse one, actually, since if some poor idiot set SSLv2_METHOD and then we unconditionally set SSL.OP_NO_SSLv2 wouldn't that mean "no connections ever"?)

If we're going to choose between the two of those, I would rather just bump endpoints users to TLSv1, since it's better security and there's no client I know of that can't handle it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

> But if upgrading the protocol version for users of SSL endpoints to TLSv1 is an incompatible change, then wouldn't upgrading the protocol version for OpenSSLCertificateOptions users who explicitly specified SSLv23_METHOD to SSL version 3 also be an incompatible change?

TLSv1 is more incompatible than SSLv3.  At some point it would be good to pick an objective metric to inform us whether we're happy to break users depending on SSLv3, but I'm confident that for now we should at least be happy to break users depending on SSLv2 (because SSLv2 is totally broken all by itself, because many installations of OpenSSL in the wild now do not even support SSLv2, because basically everyone can at least silently upgrade to SSLv3).

Also, we already did this to users once when we added `OP_NO_SSLv2` to `DefaultOpenSSLContextFactory`.  Not that two wrongs make a right, but there is precedent and I think it was right last time.

> (Perhaps a worse one, actually, since if some poor idiot set `SSLv2_METHOD` and then we unconditionally set `SSL.OP_NO_SSLv2` wouldn't that mean "no connections ever"?)

Nope.  `OP_NO_SSLv2` only changes SSLv23_METHOD behavior, nothing else.

> If we're going to choose between the two of those, I would rather just bump endpoints users to TLSv1, since it's better security and there's no client I know of that can't handle it.

I hope we'll get there someday.  However, we must consider clients *and* servers, and there are still some SSLv3-only servers out there (I would insert a list here, but it's hard to construct; the only way I know to do so is to do a lot of spidering with something that can recognize this kind of server - I know my position is weaker for not including the list, but I'm confident enough to say "trust me" here, I hope that's enough and we can move on with this very simple change).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

I'm happy to trust you, as long as we're being explicit that it's taken on faith.  Please proceed, I withdraw any objections.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)|@hynek removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

The news fragment in the patch doesn't follow the [news fragment style guide](http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_16"></a>|@tomprince commented|
|-|-|

(In [37443]) Branching to disable-sslv2-6337.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_17"></a>|@tomprince set owner to @tomprince|
|-|-|

This code looks good and implements the disucssed change. The 
[build results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/disable-sslv2-6337) look good. I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_18"></a>|@tomprince set status to `closed`|
|-|-|

(In [37445]) Merge disable-sslv2-6337: Disable SSLv2 with t.i.ssl.CertificateOptions.

Author: hynek
Reviewers: tom.prince
Fixes: [#6337](https://github.com/twisted/twisted/issues/6337)

twisted.internet.ssl.CertificateOptions now disables SSLv2 if SSLv23 is
selected, allowing only SSLv3 and TLSv1.
